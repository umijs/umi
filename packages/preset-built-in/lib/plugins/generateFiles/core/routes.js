"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _core() {
  const data = require("@umijs/core");

  _core = function _core() {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@umijs/utils");

  _utils = function _utils() {
    return data;
  };

  return data;
}

function _fs() {
  const data = require("fs");

  _fs = function _fs() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const ROUTE_FILE_EXT_LIST = ['.js', '.jsx', '.tsx'];

function _default(api) {
  const cwd = api.cwd,
        Mustache = api.utils.Mustache;
  api.onGenerateFiles( /*#__PURE__*/function () {
    var _ref = _asyncToGenerator(function* (args) {
      if ( // conventional routes
      !api.config.routes && // from watch
      args.files.length && // no files is valid route component
      !args.files.some(({
        path
      }) => {
        return path.startsWith(api.paths.absPagesPath) && ROUTE_FILE_EXT_LIST.includes((0, _path().extname)(path));
      })) {
        return;
      }

      api.logger.debug('generate core/routes.ts');
      const routesTpl = (0, _fs().readFileSync)((0, _path().join)(__dirname, 'routes.tpl'), 'utf-8');
      const routes = yield api.getRoutes();
      api.writeTmpFile({
        path: 'core/routes.ts',
        content: Mustache.render(routesTpl, {
          routes: new (_core().Route)().getJSON({
            routes,
            config: api.config,
            cwd
          }),
          runtimePath: _constants.runtimePath,
          config: api.config,
          loadingComponent: api.config.dynamicImport && api.config.dynamicImport.loading && (0, _utils().winPath)(api.config.dynamicImport.loading)
        })
      });
    });

    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }()); // 这个加进去会导致 patchRoutes 在最初就执行，但期望的是在 render 后执行
  // 所以先不加
  // api.addUmiExports(() => {
  //   return {
  //     specifiers: ['routes'],
  //     source: `./routes`,
  //   };
  // });
}