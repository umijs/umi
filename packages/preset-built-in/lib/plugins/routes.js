"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _core() {
  const data = require("@umijs/core");

  _core = function _core() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _default(api) {
  api.describe({
    key: 'routes',
    config: {
      schema(joi) {
        return joi.array().items(joi.object({
          path: joi.string().description('Any valid URL path'),
          component: joi.alternatives(joi.string(), joi.function()).description('A React component to render only when the location matches.'),
          wrappers: joi.array().items(joi.string()),
          redirect: joi.string().description('navigate to a new location'),
          exact: joi.boolean().description('When true, the active class/style will only be applied if the location is matched exactly.'),
          routes: joi.array().items(joi.link('...'))
        }).unknown());
      },

      onChange: api.ConfigChangeType.regenerateTmpFiles
    }
  });
  api.registerMethod({
    name: 'getRoutes',

    fn() {
      return _asyncToGenerator(function* () {
        const route = new (_core().Route)({
          onPatchRoutesBefore(args) {
            return _asyncToGenerator(function* () {
              yield api.applyPlugins({
                key: 'onPatchRoutesBefore',
                type: api.ApplyPluginsType.event,
                args
              });
            })();
          },

          onPatchRoutes(args) {
            return _asyncToGenerator(function* () {
              yield api.applyPlugins({
                key: 'onPatchRoutes',
                type: api.ApplyPluginsType.event,
                args
              });
            })();
          },

          onPatchRouteBefore(args) {
            return _asyncToGenerator(function* () {
              yield api.applyPlugins({
                key: 'onPatchRouteBefore',
                type: api.ApplyPluginsType.event,
                args
              });
            })();
          },

          onPatchRoute(args) {
            return _asyncToGenerator(function* () {
              yield api.applyPlugins({
                key: 'onPatchRoute',
                type: api.ApplyPluginsType.event,
                args
              });
            })();
          }

        });
        return yield api.applyPlugins({
          key: 'modifyRoutes',
          type: api.ApplyPluginsType.modify,
          initialValue: yield route.getRoutes({
            config: api.config,
            root: api.paths.absPagesPath
          })
        });
      })();
    }

  });
}