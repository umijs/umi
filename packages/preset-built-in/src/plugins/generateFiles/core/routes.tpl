import React from 'react';
import { ApplyPluginsType{{ #config.dynamicImport }}, dynamic{{ /config.dynamicImport }} } from '{{{ runtimePath }}}';
import * as umiExports from './umiExports';
import { plugin } from './plugin';
{{ #loadingComponent }}
import LoadingComponent from '{{{ loadingComponent }}}';
{{ /loadingComponent }}

export function getRoutes() {
  const routes = {{{ routes }}};

  // allow user to extend routes
  plugin.applyPlugins({
    key: 'patchRoutes',
    type: ApplyPluginsType.event,
    args: { routes },
  });

  return routes;
}
