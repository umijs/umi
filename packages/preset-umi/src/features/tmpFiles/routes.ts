import {
  addParentRoute,
  getConfigRoutes,
  getConventionRoutes,
} from '@umijs/core';
import { lodash, resolve, tryPaths, winPath, isMonorepo } from '@umijs/utils';
import { existsSync, readFileSync } from 'fs';
import { isAbsolute, join } from 'path';
import { IApi } from '../../types';
import { getModuleExports } from './getModuleExports';

// get api routes
export async function getApiRoutes(opts: { api: IApi }) {
  const routes = getConventionRoutes({
    base: opts.api.paths.absApiRoutesPath,
    prefix: '',
  });

  function localPath(path: string) {
    if (path.charAt(0) !== '.') {
      return `./${path}`;
    }
    {
      return path;
    }
  }

  for (const id of Object.keys(routes)) {
    if (routes[id].file) {
      // TODO: cache for performance
      const file = isAbsolute(routes[id].file)
        ? routes[id].file
        : resolve.sync(localPath(routes[id].file), {
            basedir: opts.api.paths.absApiRoutesPath,
            extensions: ['.js', '.jsx', '.tsx', '.ts'],
          });
      routes[id].__content = readFileSync(file, 'utf-8');
    }
  }

  return routes;
}

// get route config
export async function getRoutes(opts: {
  api: IApi;
}): Promise<Record<string, any>> {
  let routes = null;
  if (opts.api.config.routes) {
    routes = getConfigRoutes({
      routes: opts.api.config.routes,
      onResolveComponent(component) {
        if (isAbsolute(component)) {
          return component;
        }
        if (component.startsWith('@/')) {
          component = component.replace('@/', '../');
        }
        component = winPath(
          resolve.sync(localPath(component), {
            basedir: opts.api.paths.absPagesPath,
            extensions: ['.js', '.jsx', '.tsx', '.ts', '.vue'],
          }),
        );
        component = component.replace(
          winPath(`${opts.api.paths.absSrcPath}/`),
          '@/',
        );
        return component;
      },
    });
  } else {
    routes = getConventionRoutes({
      base:
        opts.api.config.conventionRoutes?.base || opts.api.paths.absPagesPath,
      exclude: opts.api.config.conventionRoutes?.exclude,
      prefix: '',
    });
  }

  function localPath(path: string) {
    if (path.charAt(0) !== '.') {
      return `./${path}`;
    } else {
      return path;
    }
  }

  for (const id of Object.keys(routes)) {
    if (routes[id].file) {
      // TODO: cache for performance
      let file = routes[id].file;
      const basedir =
        opts.api.config.conventionRoutes?.base || opts.api.paths.absPagesPath;

      if (!isAbsolute(file)) {
        if (file.startsWith('@/')) {
          file = file.replace('@/', '../');
        }
        file = resolve.sync(localPath(file), {
          basedir,
          extensions: ['.js', '.jsx', '.tsx', '.ts', '.vue'],
        });
      }

      const isJSFile = /.[jt]sx?$/.test(file);
      routes[id].__content = readFileSync(file, 'utf-8');
      routes[id].__absFile = winPath(file);
      routes[id].__isJSFile = isJSFile;
      if (opts.api.config.ssr || opts.api.config.clientLoader) {
        routes[id].__exports =
          isJSFile && existsSync(file)
            ? await getModuleExports({
                file,
              })
            : [];
      }
      if (opts.api.config.ssr) {
        routes[id].hasServerLoader =
          routes[id].__exports.includes('serverLoader');
      }
      if (opts.api.config.clientLoader) {
        routes[id].__hasClientLoader =
          routes[id].__exports.includes('clientLoader');
        routes[id].clientLoader = `clientLoaders['${id}']`;
      }
    }
  }

  // layout routes
  const absLayoutPath = tryPaths([
    join(opts.api.paths.absSrcPath, 'layouts/index.tsx'),
    join(opts.api.paths.absSrcPath, 'layouts/index.vue'),
    join(opts.api.paths.absSrcPath, 'layouts/index.jsx'),
    join(opts.api.paths.absSrcPath, 'layouts/index.js'),
  ]);

  const layouts = await opts.api.applyPlugins({
    key: 'addLayouts',
    initialValue: [
      absLayoutPath && {
        id: '@@/global-layout',
        file: winPath(absLayoutPath),
        test(route: any) {
          return route.layout !== false;
        },
      },
    ].filter(Boolean),
  });
  for (const layout of layouts) {
    addParentRoute({
      addToAll: true,
      target: {
        id: layout.id,
        path: '/',
        file: layout.file,
        parentId: undefined,
        absPath: '/',
        isLayout: true,
      },
      routes,
      test: layout.test,
    });
  }

  // patch routes
  for (const id of Object.keys(routes)) {
    await opts.api.applyPlugins({
      key: 'onPatchRoute',
      args: {
        route: routes[id],
      },
    });
  }

  routes = await opts.api.applyPlugins({
    key: 'modifyRoutes',
    initialValue: routes,
  });

  return routes;
}

export async function getRouteComponents(opts: {
  routes: Record<string, any>;
  prefix: string;
  api: IApi;
}) {
  const imports = Object.keys(opts.routes)
    .map((key) => {
      const useSuspense = opts.api.appData.framework === 'react' ? true : false; // opts.api.appData.react.version.startsWith('18.');
      const route = opts.routes[key];
      if (!route.file) {
        // 测试环境还不支持 import ，所以用 require
        if (process.env.NODE_ENV === 'test') {
          return `'${key}': require( './EmptyRoute').default,`;
        }
        return useSuspense
          ? `'${key}': React.lazy(() => import( './EmptyRoute')),`
          : `'${key}': () => import( './EmptyRoute'),`;
      }
      if (route.hasClientLoader) {
        route.file = join(
          opts.api.paths.absTmpPath,
          'pages',
          route.id.replace(/[\/\-]/g, '_') + '.js',
        );
      }
      // e.g.
      // component: () => <h1>foo</h1>
      // component: (() => () => <h1>foo</h1>)()
      if (route.file.startsWith('(')) {
        return useSuspense
          ? // Compatible with none default route exports
            // e.g. https://github.com/umijs/umi/blob/0d40a07bf28b0760096cbe2f22da4d639645b937/packages/plugins/src/qiankun/master.ts#L55
            `'${key}': React.lazy(
              () => Promise.resolve(${route.file}).then(e => e?.default ? e : ({ default: e }))
            ),`
          : `'${key}': () => Promise.resolve(${route.file}),`;
      }

      const path =
        isAbsolute(route.file) || route.file.startsWith('@/')
          ? route.file
          : `${opts.prefix}${route.file}`;

      const webpackChunkName = componentToChunkName(path, opts.api.cwd);

      // 测试环境还不支持 import ，所以用 require
      if (process.env.NODE_ENV === 'test') {
        return `'${key}': require('${winPath(path)}').default,`;
      }

      return useSuspense
        ? `'${key}': React.lazy(() => import(/* webpackChunkName: "${webpackChunkName}" */'${winPath(
            path,
          )}')),`
        : `'${key}': () => import(/* webpackChunkName: "${webpackChunkName}" */'${winPath(
            path,
          )}'),`;
    })
    .join('\n');
  return `{\n${imports}\n}`;
}

function lastSlash(str: string) {
  return str[str.length - 1] === '/' ? str : `${str}/`;
}

function getProjectRootCwd(cwd: string) {
  const splittedCwd = cwd.split('/');

  // try to find root cwd for monorepo project, only support >= 3 level depth
  for (let level = -1; level >= -3; level -= 1) {
    const rootCwd = splittedCwd.slice(0, level).join('/');

    // break if no parent dir
    if (!rootCwd) break;

    // check monorepo for parent dir
    if (isMonorepo({ root: rootCwd })) {
      return rootCwd;
    }
  }

  return cwd;
}

/**
 *
 * transform component into webpack chunkName
 * @export
 * @param {string} component component path
 * @param {string} [cwd] current root path
 * @return {*}  {string}
 */
export function componentToChunkName(
  component: string,
  cwd: string = '/',
): string {
  cwd = winPath(cwd);

  return typeof component === 'string'
    ? component
        .replace(
          new RegExp(
            `^(${
              // match app cwd first
              lodash.escapeRegExp(lastSlash(cwd))
            }|${
              // then try to match monorepo root cwd, because route files may be in root node_modules (such as dumi)
              lodash.escapeRegExp(lastSlash(getProjectRootCwd(cwd)))
            })`,
          ),
          '',
        )
        .replace(/^.(\/|\\)/, '')
        .replace(/(\/|\\)/g, '__')
        // 转换 tnpm node_modules 目录中的 @ 符号，它在 URL 上会被转义，可能导致 CDN 托管失败
        .replace(/@/g, '_')
        .replace(/\.jsx?$/, '')
        .replace(/\.tsx?$/, '')
        .replace(/\.vue?$/, '')
        .replace(/^src__/, '')
        .replace(/\.\.__/g, '')
        // 约定式路由的 [ 会导致 webpack 的 code splitting 失败
        // ref: https://github.com/umijs/umi/issues/4155
        .replace(/[\[\]]/g, '')
        // node_modules 文件名在 gh-pages 是默认忽略的，会导致访问 404
        .replace(/^node_modules__/, 'nm__')
        // 插件层的文件也可能是路由组件，比如 plugin-layout 插件
        .replace(/^.umi-production__/, 't__')
        // 避免产出隐藏文件（比如 .dumi/theme）下的路由组件
        .replace(/^\./, '')
        .replace(/^pages__/, 'p__')
    : '';
}
