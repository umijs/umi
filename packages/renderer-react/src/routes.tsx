// @ts-ignore
import React, { useMemo } from 'react';
import { generatePath, Navigate, useParams, Outlet } from 'react-router-dom';
import { RouteContext, useRouteData } from './routeContext';
import { IClientRoute, IRoute, IRoutesById } from './types';
import { useAppData } from './appContext';

export function createClientRoutes(opts: {
  routesById: IRoutesById;
  routeComponents: Record<string, any>;
  parentId?: string;
  loadingComponent?: React.ReactNode;
  reactRouter5Compat?: boolean;
}) {
  const { routesById, parentId, routeComponents } = opts;
  return Object.keys(routesById)
    .filter((id) => routesById[id].parentId === parentId)
    .map((id) => {
      const route = createClientRoute({
        route: routesById[id],
        routeComponent: routeComponents[id],
        loadingComponent: opts.loadingComponent,
        reactRouter5Compat: opts.reactRouter5Compat,
        ...(opts.reactRouter5Compat && {
          // TODO: 这个不准，没考虑 patchClientRoutes 的场景
          hasChildren:
            Object.keys(routesById).filter(
              (rid) => routesById[rid].parentId === id,
            ).length > 0,
        }),
      });
      const children = createClientRoutes({
        routesById,
        routeComponents,
        parentId: route.id,
        loadingComponent: opts.loadingComponent,
        reactRouter5Compat: opts.reactRouter5Compat,
      });
      if (children.length > 0) {
        route.children = children;
        // TODO: remove me
        // compatible with @ant-design/pro-layout
        route.routes = children;
      }
      return route;
    });
}

function NavigateWithParams(props: { to: string }) {
  const params = useParams();
  const propsWithParams = {
    ...props,
    to: generatePath(props.to, params),
  };
  return <Navigate replace={true} {...propsWithParams} />;
}

function createClientRoute(opts: {
  route: IRoute;
  routeComponent: any;
  loadingComponent?: React.ReactNode;
  hasChildren?: boolean;
  reactRouter5Compat?: boolean;
}): IClientRoute {
  const { route } = opts;
  const { redirect, ...props } = route;
  const Remote = opts.reactRouter5Compat
    ? RemoteComponentReactRouter5
    : RemoteComponent;
  return {
    element: redirect ? (
      <NavigateWithParams to={redirect} />
    ) : (
      <RouteContext.Provider
        value={{
          route: opts.route,
        }}
      >
        <Remote
          loader={React.memo(opts.routeComponent)}
          loadingComponent={opts.loadingComponent || DefaultLoading}
          hasChildren={opts.hasChildren}
        />
      </RouteContext.Provider>
    ),
    ...props,
  };
}

function DefaultLoading() {
  return <div />;
}

function RemoteComponentReactRouter5(props: any) {
  const { route } = useRouteData();
  const { history, clientRoutes } = useAppData();
  const params = useParams();
  const match = {
    params,
    isExact: true,
    path: route.path,
    url: history.location.pathname,
  };

  // staticContext 没有兼容 好像没看到对应的兼容写法
  const Component = props.loader;

  return (
    <React.Suspense fallback={<props.loadingComponent />}>
      <Component
        location={history.location}
        match={match}
        history={history}
        params={params}
        route={route}
        routes={clientRoutes}
      >
        {props.hasChildren && <Outlet />}
      </Component>
    </React.Suspense>
  );
}

function RemoteComponent(props: any) {
  const Component = props.loader;
  return (
    <React.Suspense fallback={<props.loadingComponent />}>
      <Component />
    </React.Suspense>
  );
}
