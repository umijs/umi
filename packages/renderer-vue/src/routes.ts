import { IRoute, IRoutesById } from './types';

export function createClientRoutes(opts: {
  routesById: IRoutesById;
  routeComponents: Record<string, any>;
  parentId?: string;
}) {
  const { routesById, parentId, routeComponents } = opts;
  return Object.keys(routesById)
    .filter((id) => routesById[id].parentId === parentId)
    .map((id) => {
      const route = createClientRoute({
        route: routesById[id],
        routeComponent: routeComponents[id],
        parentId,
      });

      const children = createClientRoutes({
        routesById,
        routeComponents,
        parentId: route.id,
      });

      if (children.length > 0) {
        // @ts-ignore
        route.children = children;
      }

      delete route.id;

      return route;
    });
}

export function createClientRoute(opts: {
  route: IRoute;
  routeComponent: any;
  parentId?: string;
}) {
  const { route } = opts;
  const { id, path, redirect, ...other } = route;

  if (redirect) {
    return {
      ...other,
      id,
      path,
      redirect,
    };
  }

  const item: Record<string, any> = {
    ...other,
    id,
    component: opts.routeComponent,
    path,
  };

  if (route.parentId === undefined && path && !path.startsWith('/')) {
    // fix Uncaught (in promise) Error: Route paths should start with a "/": "users" should be "/users".
    item.path = `/${path}`;
  }

  return item;
}
