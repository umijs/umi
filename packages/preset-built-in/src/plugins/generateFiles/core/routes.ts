import { Route } from '@umijs/core';
import { IApi } from '@umijs/types';
import { winPath } from '@umijs/utils';
import { readFileSync } from 'fs';
import { extname, join } from 'path';
import { runtimePath } from '../constants';

const ROUTE_FILE_EXT_LIST = ['.js', '.jsx', '.tsx'];

export default function (api: IApi) {
  const {
    cwd,
    utils: { Mustache },
  } = api;

  api.onGenerateFiles(async (args) => {
    if (
      // conventional routes
      !api.config.routes &&
      // from watch
      args.files.length &&
      // no files is valid route component
      !args.files.some(({ path }) => {
        return (
          path.startsWith(api.paths.absPagesPath!) &&
          ROUTE_FILE_EXT_LIST.includes(extname(path))
        );
      })
    ) {
      return;
    }

    api.logger.debug('generate core/routes.ts');
    const routesTpl = readFileSync(join(__dirname, 'routes.tpl'), 'utf-8');
    const routes = await api.getRoutes();
    api.writeTmpFile({
      path: 'core/routes.ts',
      content: Mustache.render(routesTpl, {
        routes: new Route().getJSON({ routes, config: api.config, cwd }),
        runtimePath,
        config: api.config,
        loadingComponent:
          api.config.dynamicImport &&
          api.config.dynamicImport.loading &&
          winPath(api.config.dynamicImport.loading),
      }),
    });
  });

  // 这个加进去会导致 patchRoutes 在最初就执行，但期望的是在 render 后执行
  // 所以先不加
  // api.addUmiExports(() => {
  //   return {
  //     specifiers: ['routes'],
  //     source: `./routes`,
  //   };
  // });
}
