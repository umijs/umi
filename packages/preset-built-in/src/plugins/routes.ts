import { Route } from '@umijs/core';
import { IApi } from '@umijs/types';

export default function (api: IApi) {
  api.describe({
    key: 'routes',
    config: {
      schema(joi) {
        return joi.array().items(
          joi
            .object({
              path: joi.string().description('Any valid URL path'),
              component: joi
                .alternatives(joi.string(), joi.function())
                .description(
                  'A React component to render only when the location matches.',
                ),
              wrappers: joi.array().items(joi.string()),
              redirect: joi.string().description('navigate to a new location'),
              exact: joi
                .boolean()
                .description(
                  'When true, the active class/style will only be applied if the location is matched exactly.',
                ),
              routes: joi.array().items(joi.link('...')),
            })
            .unknown(),
        );
      },
      onChange: api.ConfigChangeType.regenerateTmpFiles,
    },
  });

  api.registerMethod({
    name: 'getRoutes',
    async fn() {
      const route = new Route({
        async onPatchRoutesBefore(args: object) {
          await api.applyPlugins({
            key: 'onPatchRoutesBefore',
            type: api.ApplyPluginsType.event,
            args,
          });
        },
        async onPatchRoutes(args: object) {
          await api.applyPlugins({
            key: 'onPatchRoutes',
            type: api.ApplyPluginsType.event,
            args,
          });
        },
        async onPatchRouteBefore(args: object) {
          await api.applyPlugins({
            key: 'onPatchRouteBefore',
            type: api.ApplyPluginsType.event,
            args,
          });
        },
        async onPatchRoute(args: object) {
          await api.applyPlugins({
            key: 'onPatchRoute',
            type: api.ApplyPluginsType.event,
            args,
          });
        },
      });
      return await api.applyPlugins({
        key: 'modifyRoutes',
        type: api.ApplyPluginsType.modify,
        initialValue: await route.getRoutes({
          config: api.config,
          root: api.paths.absPagesPath!,
        }),
      });
    },
  });
}
