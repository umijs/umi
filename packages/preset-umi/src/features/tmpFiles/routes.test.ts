import { join } from 'path';
import { componentToChunkName, getRoutes } from './routes';

const fixtures = join(__dirname, './fixtures/getRoutes');
const ABSOLUTE_PAGE_PATH = join(
  __dirname,
  './fixtures/getRoutes/pages/absolute.tsx',
);

test('getRoutes', async () => {
  const routes = await getRoutes({
    api: {
      paths: {
        absSrcPath: fixtures,
        absPagesPath: join(fixtures, 'pages'),
      },
      config: {
        routes: [
          {
            id: 'login',
            path: '/login',
            layout: false,
            component: 'users/Login',
          },
          {
            path: '*',
            component: '404',
          },
          {
            path: '/hello',
            component: '@/pages/hello',
          },
          {
            path: '/users',
            component: '@/pages/users/index',
            routes: [
              {
                path: '/users/foo',
                component: './users/Foo',
              },
            ],
          },
          {
            path: '/absolute',
            component: ABSOLUTE_PAGE_PATH,
          },
        ],
      },
      applyPlugins(opts: any) {
        return opts?.initialValue;
      },
    } as any,
  });

  // login 页不包含布局
  expect(routes[1].file).toBe('@/pages/users/Login/index.tsx');
  expect(routes[1].parentId).toBe(undefined);

  // @@/global-layout
  expect(routes['@@/global-layout'].file).toContain('layouts/index.tsx');
  expect(routes['@@/global-layout'].parentId).toBe(undefined);
  expect(routes['@@/global-layout'].isLayout).toBe(true);

  // 普通页面增加布局
  expect(routes[2].file).toBe('@/pages/404.tsx');
  expect(routes[2].parentId).toBe('@@/global-layout');

  // vue 文件测试
  expect(routes[3].file).toBe('@/pages/hello/index.vue');
  expect(routes[3].parentId).toBe('@@/global-layout');

  // resolve absolute path
  expect(routes[6].file).toBe(ABSOLUTE_PAGE_PATH);
  delete routes[6].file;

  // __absFile 是具体的路径, 快照测试通不过
  Object.keys(routes).forEach((id) => {
    delete routes[id].__absFile;
  });

  // 覆写 layout 的绝对路径地址，用于保持快照稳定
  routes['@@/global-layout'].file = '@/layouts/index.tsx';

  expect(routes).toMatchSnapshot();
});

test('componentToChunkName normal', () => {
  expect(componentToChunkName('@/pages/index.ts')).toEqual('p__index');

  expect(componentToChunkName('@/pages/index.tsx')).toEqual('p__index');

  expect(componentToChunkName('@/pages/index.jsx')).toEqual('p__index');

  expect(componentToChunkName('@/pages/index.js')).toEqual('p__index');

  expect(componentToChunkName('@/pages/users/[id].ts')).toEqual('p__users__id');

  expect(componentToChunkName('@/pages/users/[id].tsx')).toEqual(
    'p__users__id',
  );

  expect(componentToChunkName('@/pages/users/[id].js')).toEqual('p__users__id');

  expect(componentToChunkName('@/pages/users/[id].jsx')).toEqual(
    'p__users__id',
  );

  expect(componentToChunkName('@/pages/users/[id].vue')).toEqual(
    'p__users__id',
  );

  expect(componentToChunkName('@/components/404/index.tsx')).toEqual(
    'components__404__index',
  );

  expect(componentToChunkName('@/layouts/index.tsx')).toEqual('layouts__index');

  expect(
    componentToChunkName('@/.umi-production/plugin-layout/Layout.tsx'),
  ).toEqual('t__plugin-layout__Layout');
});

test('componentToChunkName cwd', () => {
  expect(
    componentToChunkName('/users/test/pages/users/[id].jsx', '/users/test'),
  ).toEqual('p__users__id');

  expect(
    componentToChunkName('/users/test/pages/users/[id].vue', '/users/test'),
  ).toEqual('p__users__id');
});

test('componentToChunkName cwd escape char', () => {
  expect(
    componentToChunkName('/users/c++/pages/users/[id].tsx', '/users/c++'),
  ).toEqual('p__users__id');
});
