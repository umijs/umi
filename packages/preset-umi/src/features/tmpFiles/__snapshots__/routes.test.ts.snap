// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRoutes 1`] = `
Object {
  "1": Object {
    "__content": "",
    "__isJSFile": true,
    "absPath": "/login",
    "file": "@/pages/users/Login/index.tsx",
    "id": "1",
    "layout": false,
    "parentId": undefined,
    "path": "/login",
  },
  "2": Object {
    "__content": "",
    "__isJSFile": true,
    "absPath": "/*",
    "file": "@/pages/404.tsx",
    "id": "2",
    "parentId": "@@/global-layout",
    "path": "*",
  },
  "3": Object {
    "__content": "",
    "__isJSFile": false,
    "absPath": "/hello",
    "file": "@/pages/hello/index.vue",
    "id": "3",
    "parentId": "@@/global-layout",
    "path": "/hello",
  },
  "4": Object {
    "__content": "",
    "__isJSFile": false,
    "absPath": "/users",
    "file": "@/pages/users/index.vue",
    "id": "4",
    "parentId": "@@/global-layout",
    "path": "/users",
  },
  "5": Object {
    "__content": "",
    "__isJSFile": true,
    "absPath": "/users/foo",
    "file": "@/pages/users/Foo.tsx",
    "id": "5",
    "parentId": "4",
    "path": "/users/foo",
  },
  "6": Object {
    "__content": "",
    "__isJSFile": true,
    "absPath": "/absolute",
    "id": "6",
    "parentId": "@@/global-layout",
    "path": "/absolute",
  },
  "@@/global-layout": Object {
    "absPath": "/",
    "file": "@/layouts/index.tsx",
    "id": "@@/global-layout",
    "isLayout": true,
    "parentId": undefined,
    "path": "/",
  },
}
`;
